# sanger-tol/metagenomeassembly: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [Aria2](https://aria2.github.io/)

  > aria2 [WWW Document], n.d. URL https://aria2.github.io/ (accessed 3.7.25).

- [bin3C](https://github.com/cerebis/bin3C)

  > DeMaere, M.Z., Darling, A.E., 2019. bin3C: exploiting Hi-C sequencing data to accurately resolve metagenome-assembled genomes. Genome Biology 20, 46. https://doi.org/10.1186/s13059-019-1643-1

- [bwa-mem2](https://github.com/bwa-mem2/bwa-mem2)

  > Vasimuddin, Md., Misra, S., Li, H., Aluru, S., 2019. Efficient Architecture-Aware Acceleration of BWA-MEM for Multicore Systems, in: 2019 IEEE International Parallel and Distributed Processing Symposium (IPDPS). Presented at the 2019 IEEE International Parallel and Distributed Processing Symposium (IPDPS), pp. 314–324. https://doi.org/10.1109/IPDPS.2019.00041

- [CheckM2](https://github.com/chklovski/CheckM2)

  > Chklovski, A., Parks, D.H., Woodcroft, B.J., Tyson, G.W., 2023. CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning. Nat Methods 20, 1203–1212. https://doi.org/10.1038/s41592-023-01940-w

- [Coverm](https://github.com/wwood/CoverM)

  > Aroney, S.T.N., Newell, R.J.P., Nissen, J.N., Camargo, A.P., Tyson, G.W., Woodcroft, B.J., 2025. CoverM: Read alignment statistics for metagenomics. https://doi.org/10.48550/arXiv.2501.11217

- [DAS_Tool](https://github.com/cmks/DAS_Tool)

  > Sieber, C.M.K., Probst, A.J., Sharrar, A., Thomas, B.C., Hess, M., Tringe, S.G., Banfield, J.F., 2018. Recovery of genomes from metagenomes via a dereplication, aggregation and scoring strategy. Nat Microbiol 3, 836–843. https://doi.org/10.1038/s41564-018-0171-1

- [GNU Awk](https://www.gnu.org/software/gawk/manual/gawk.html)

  > Robbins, A., 2003. GAWK: effective AWK programming : a user’s guide for GNU Awk, Ed. 3. ed. Free Software Foundation, Boston, MA.

- [GNU Core utilities](https://www.gnu.org/software/coreutils/)

  > Coreutils - GNU core utilities [WWW Document], n.d. URL https://www.gnu.org/software/coreutils/ (accessed 3.10.25).

- [GTDB-Tk](https://github.com/Ecogenomics/GTDBTk)

  > Chaumeil, P.-A., Mussig, A.J., Hugenholtz, P., Parks, D.H., 2022. GTDB-Tk v2: memory friendly classification with the genome taxonomy database. Bioinformatics 38, 5315–5316. https://doi.org/10.1093/bioinformatics/btac672

- [hmmer](http://hmmer.org/)

  > Eddy, S.R., 2011. Accelerated Profile HMM Searches. PLOS Computational Biology 7, e1002195. https://doi.org/10.1371/journal.pcbi.1002195

- [Infernal](http://eddylab.org/infernal/)

  > Nawrocki, E.P., Eddy, S.R., 2013. Infernal 1.1: 100-fold faster RNA homology searches. Bioinformatics 29, 2933–2935. https://doi.org/10.1093/bioinformatics/btt509

- [MAGScoT](https://github.com/ikmb/MAGScoT)

  > Rühlemann, M.C., Wacker, E.M., Ellinghaus, D., Franke, A., 2022. MAGScoT: a fast, lightweight and accurate bin-refinement tool. Bioinformatics 38, 5430–5433. https://doi.org/10.1093/bioinformatics/btac694

- [MaxBin2](https://sourceforge.net/projects/maxbin2/)

  > Wu, Y.-W., Simmons, B.A., Singer, S.W., 2016. MaxBin 2.0: an automated binning algorithm to recover genomes from multiple metagenomic datasets. Bioinformatics 32, 605–607. https://doi.org/10.1093/bioinformatics/btv638

- [MetaBat2](https://bitbucket.org/berkeleylab/metabat)

  > Kang, D.D., Li, F., Kirton, E., Thomas, A., Egan, R., An, H., Wang, Z., 2019. MetaBAT 2: an adaptive binning algorithm for robust and efficient genome reconstruction from metagenome assemblies. PeerJ 7, e7359. https://doi.org/10.7717/peerj.7359

- [metaMDBG](https://github.com/GaetanBenoitDev/metaMDBG)

  > Benoit, G., Raguideau, S., James, R., Phillippy, A.M., Chikhi, R., Quince, C., 2024. High-quality metagenome assembly from long accurate reads with metaMDBG. Nat Biotechnol 42, 1378–1383. https://doi.org/10.1038/s41587-023-01983-6

- [Metator](https://github.com/koszullab/metaTOR/)

  > Baudry, L., Foutel-Rodier, T., Thierry, A., Koszul, R., Marbouty, M., 2019. MetaTOR: A Computational Pipeline to Recover High-Quality Metagenomic Bins From Mammalian Gut Proximity-Ligation (meta3C) Libraries. Front. Genet. 10. https://doi.org/10.3389/fgene.2019.00753

- [minimap2](https://github.com/lh3/minimap2)

  > Li, H., 2021. New strategies to improve minimap2 alignment accuracy. Bioinformatics 37, 4572–4574. https://doi.org/10.1093/bioinformatics/btab705

- [Pyrodigal](https://pyrodigal.readthedocs.io/)

  > Larralde, M., 2022. Pyrodigal: Python bindings and interface to Prodigal, an efficient method for gene prediction in prokaryotes. Journal of Open Source Software 7, 4296. https://doi.org/10.21105/joss.04296

- [R](https://www.r-project.org/)

  > R Core Team, 2025. R: A Language and Environment for Statistical Computing. R Foundation for Statistical Computing, Vienna, Austria.

- [samtools](https://www.htslib.org/)

  > Danecek, P., Bonfield, J.K., Liddle, J., Marshall, J., Ohan, V., Pollard, M.O., Whitwham, A., Keane, T., McCarthy, S.A., Davies, R.M., Li, H., 2021. Twelve years of SAMtools and BCFtools. Gigascience 10, giab008. https://doi.org/10.1093/gigascience/giab008

- [SeqKit](https://bioinf.shenwei.me/seqkit/)

  > Shen, W., Sipos, B., Zhao, L., 2024. SeqKit2: A Swiss army knife for sequence and alignment processing. iMeta 3, e191. https://doi.org/10.1002/imt2.191

- [staden-io-lib](https://github.com/jkbonfield/io_lib)

  > Bonfield, J., 2024. jkbonfield/io_lib (https://github.com/jkbonfield/io_lib)

- [TaxonKit](https://bioinf.shenwei.me/taxonkit/)

  > Shen, W., Ren, H., 2021. TaxonKit: A practical and efficient NCBI taxonomy toolkit. Journal of Genetics and Genomics, Special issue on Microbiome 48, 844–850. https://doi.org/10.1016/j.jgg.2021.03.006

- [tidyverse](https://www.tidyverse.org/)

  > Wickham, H., Averick, M., Bryan, J., Chang, W., McGowan, L.D., François, R., Grolemund, G., Hayes, A., Henry, L., Hester, J., Kuhn, M., Pedersen, T.L., Miller, E., Bache, S.M., Müller, K., Ooms, J., Robinson, D., Seidel, D.P., Spinu, V., Takahashi, K., Vaughan, D., Wilke, C., Woo, K., Yutani, H., 2019. Welcome to the tidyverse. Journal of Open Source Software 4, 1686. https://doi.org/10.21105/joss.01686

- [tRNAScan-SE](https://github.com/UCSC-LoweLab/tRNAscan-SE)

  > Chan, P.P., Lin, B.Y., Mak, A.J., Lowe, T.M., 2021. tRNAscan-SE 2.0: improved detection and functional classification of transfer RNA genes. Nucleic Acids Research 49, 9077–9096. https://doi.org/10.1093/nar/gkab688

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
